File: frontend/src/features/automations/tree.ts
// Arbre de règle d'un automate — miroir TS du schéma backend automations/tree.py.
// Blocs récursifs : filtre (arbre ET/OU imbriqué) → appels nommés → blocs enfants.

export interface TreeHeader {
  name: string
  value?: string | null
  secret_ref?: string | null
  value_prefix?: string
  required?: boolean
  enabled?: boolean
}

export interface TreeCall {
  name: string
  url: string
  http_method: string
  body_template?: string | null
  headers: TreeHeader[]
  contract_ref?: string | null
  operation_id?: string | null
}

export interface TreeFilterLeaf {
  url: string
  http_method: string
  body?: string | null
  jsonpath: string
  operator: string
  expected?: string | null
  headers: TreeHeader[]
  contract_ref?: string | null
  operation_id?: string | null
}

export interface TreeFilterGroup {
  op: 'and' | 'or'
  items: TreeFilterNode[]
}

export type TreeFilterNode = TreeFilterGroup | TreeFilterLeaf

export interface TreeBlock {
  label: string
  filter: TreeFilterNode | null
  calls: TreeCall[]
  blocks: TreeBlock[]
}

export interface RuleTree {
  version: 1
  blocks: TreeBlock[]
}

export const OPERATORS = ['exists', 'not_exists', 'equals', 'not_equals'] as const

export function isGroup(node: TreeFilterNode): node is TreeFilterGroup {
  return (node as TreeFilterGroup).op !== undefined
}

export function emptyTree(): RuleTree {
  return { version: 1, blocks: [] }
}

export function newBlock(): TreeBlock {
  return { label: '', filter: null, calls: [], blocks: [] }
}

export function newCall(existingNames: string[]): TreeCall {
  let i = existingNames.length + 1
  let name = `call${i}`
  while (existingNames.includes(name)) name = `call${++i}`
  return { name, url: '', http_method: 'POST', body_template: null, headers: [] }
}

export function newLeaf(): TreeFilterLeaf {
  return {
    url: '',
    http_method: 'GET',
    jsonpath: '',
    operator: 'exists',
    expected: null,
    headers: [],
  }
}

/** Fusionne les en-têtes d'auth d'une opération dans une liste (sans doublon de nom). */
export function mergeAuthHeaders(
  existing: TreeHeader[],
  auth: { header: string; value_prefix: string }[] | undefined,
): TreeHeader[] {
  if (!auth?.length) return existing
  const names = new Set(existing.map((h) => h.name.toLowerCase()))
  const add = auth
    .filter((a) => !names.has(a.header.toLowerCase()))
    .map<TreeHeader>((a) => ({
      name: a.header,
      value: null,
      secret_ref: null,
      value_prefix: a.value_prefix,
      required: true,
      enabled: true,
    }))
  return add.length ? [...existing, ...add] : existing
}

export function newGroup(first: TreeFilterNode): TreeFilterGroup {
  return { op: 'and', items: [first] }
}

/** Tous les noms d'appels de l'arbre (parcours en profondeur). */
export function collectCallNames(blocks: TreeBlock[]): string[] {
  const out: string[] = []
  for (const b of blocks) {
    for (const c of b.calls) out.push(c.name)
    out.push(...collectCallNames(b.blocks))
  }
  return out
}

/** Compteurs pour la ligne de liste ({{blocks}} bloc(s), {{calls}} appel(s)). */
export function treeSummary(tree: RuleTree | null | undefined): { blocks: number; calls: number } {
  function walk(blocks: TreeBlock[]): { blocks: number; calls: number } {
    let nb = 0
    let nc = 0
    for (const b of blocks) {
      nb += 1
      nc += b.calls.length
      const sub = walk(b.blocks)
      nb += sub.blocks
      nc += sub.calls
    }
    return { blocks: nb, calls: nc }
  }
  return walk(tree?.blocks ?? [])
}

/** Variables {var} référencées dans l'arbre (pour les valeurs d'exemple des tests). */
export function collectUsedVariables(tree: RuleTree): string[] {
  const found = new Set<string>()
  const scan = (s: string | null | undefined) => {
    // Même charset que le moteur de template backend (_VAR_RE) : évite de
    // capturer des accolades de JSON comme des variables.
    for (const m of (s ?? '').matchAll(/\{([a-zA-Z0-9_.]+)\}/g)) found.add(m[1])
  }
  function walkFilter(node: TreeFilterNode | null) {
    if (!node) return
    if (isGroup(node)) {
      node.items.forEach(walkFilter)
      return
    }
    scan(node.url)
    scan(node.body)
    scan(node.jsonpath)
    scan(node.expected)
  }
  function walkBlocks(blocks: TreeBlock[]) {
    for (const b of blocks) {
      walkFilter(b.filter)
      for (const c of b.calls) {
        scan(c.url)
        scan(c.body_template)
      }
      walkBlocks(b.blocks)
    }
  }
  walkBlocks(tree.blocks)
  return [...found]
}

File: backend/src/portal/automations/tree.py
"""Arbre de règle d'un automate (champ JSONB `automation.tree`).

Une règle = une liste de blocs **récursifs**. Chaque bloc porte un filtre
(arbre ET/OU imbriqué de feuilles « appel + évaluation JSONPath »), des appels
nommés, et des blocs enfants. Parcours en profondeur : filtre → s'il passe,
appels puis enfants ; sinon le sous-arbre est sauté et le bloc frère suivant
continue. La réponse JSON de chaque appel est rangée dans un dictionnaire sous
son `name` : les templates aval y accèdent par `{<name>.chemin.vers.champ}`.

Le schéma est LA validation unique : routes API et primitives MCP valident via
`RuleTree.model_validate`, le runner re-parse le JSONB au moment d'exécuter.
"""

from __future__ import annotations

import re
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from .filter_eval import OPERATORS

HTTP_METHODS = ("GET", "POST", "PUT", "PATCH", "DELETE")

# Nom d'un appel : racine de variable de template ({name.champ...}) → pas de point.
_CALL_NAME_RE = re.compile(r"^[a-zA-Z_][a-zA-Z0-9_]{0,63}$")


class TreeHeader(BaseModel):
    """En-tête d'un appel/filtre : `value` XOR `secret_ref` (les deux None = stub).

    `secret_ref` = référence vault résolue à l'exécution (`${system://slug}` ou
    `${vault://…}`) ; `value_prefix` est concaténé devant (ex. « Bearer »).
    """

    model_config = ConfigDict(extra="forbid")

    name: str
    value: str | None = None
    secret_ref: str | None = None
    value_prefix: str = ""
    required: bool = False
    enabled: bool = True

    @model_validator(mode="after")
    def _value_xor_secret(self) -> TreeHeader:
        if self.value is not None and self.secret_ref is not None:
            raise ValueError(f"en-tête {self.name!r} : value et secret_ref exclusifs")
        return self


class TreeCall(BaseModel):
    """Appel HTTP nommé ; sa réponse JSON est exposée sous `name` aux blocs aval."""

    model_config = ConfigDict(extra="forbid")

    name: str
    url: str
    http_method: str
    body_template: str | None = None
    # En-têtes propres à l'appel (pré-remplis depuis l'opération du contrat).
    headers: list[TreeHeader] = Field(default_factory=list)
    # Métadonnées d'édition (opération de contrat choisie dans l'IHM).
    contract_ref: str | None = None
    operation_id: str | None = None

    @field_validator("name")
    @classmethod
    def _name_is_template_root(cls, v: str) -> str:
        if not _CALL_NAME_RE.fullmatch(v):
            raise ValueError(
                f"nom d'appel invalide : {v!r} (lettres/chiffres/_, sans point, max 64)"
            )
        return v

    @field_validator("http_method")
    @classmethod
    def _method_known(cls, v: str) -> str:
        if v not in HTTP_METHODS:
            raise ValueError(f"méthode HTTP invalide : {v!r}")
        return v

    @field_validator("url")
    @classmethod
    def _url_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("url d'appel vide")
        return v


class TreeFilterLeaf(BaseModel):
    """Feuille de filtre : appel HTTP + évaluation JSONPath/opérateur sur la réponse."""

    model_config = ConfigDict(extra="forbid")

    url: str
    http_method: str = "GET"
    body: str | None = None
    jsonpath: str
    operator: str
    expected: str | None = None
    # En-têtes propres au filtre (pré-remplis depuis l'opération du contrat).
    headers: list[TreeHeader] = Field(default_factory=list)
    contract_ref: str | None = None
    operation_id: str | None = None

    @field_validator("http_method")
    @classmethod
    def _method_known(cls, v: str) -> str:
        if v not in HTTP_METHODS:
            raise ValueError(f"méthode HTTP invalide : {v!r}")
        return v

    @field_validator("operator")
    @classmethod
    def _operator_known(cls, v: str) -> str:
        if v not in OPERATORS:
            raise ValueError(f"opérateur inconnu : {v!r} (attendus : {', '.join(OPERATORS)})")
        return v

    @field_validator("url")
    @classmethod
    def _url_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("url de filtre vide")
        return v


class TreeFilterGroup(BaseModel):
    """Groupe ET/OU d'un arbre de filtres (imbrication libre)."""

    model_config = ConfigDict(extra="forbid")

    op: Literal["and", "or"]
    items: list[TreeFilterNode] = Field(min_length=1)


# Union structurelle : une feuille n'a pas de champ `op`, un groupe n'a que
# `op` + `items` — extra="forbid" rend la résolution non ambiguë.
TreeFilterNode = TreeFilterGroup | TreeFilterLeaf


class TreeBlock(BaseModel):
    """Bloc récursif : filtre (porte) → appels → blocs enfants."""

    model_config = ConfigDict(extra="forbid")

    label: str = ""
    filter: TreeFilterNode | None = None
    calls: list[TreeCall] = Field(default_factory=list)
    blocks: list[TreeBlock] = Field(default_factory=list)


class RuleTree(BaseModel):
    """Racine du champ `automation.tree`."""

    model_config = ConfigDict(extra="forbid")

    version: Literal[1] = 1
    blocks: list[TreeBlock] = Field(default_factory=list)

    @model_validator(mode="after")
    def _call_names_unique(self) -> RuleTree:
        seen: set[str] = set()
        for call in iter_calls(self.blocks):
            if call.name in seen:
                raise ValueError(f"nom d'appel dupliqué dans la règle : {call.name!r}")
            seen.add(call.name)
        return self


def iter_calls(blocks: list[TreeBlock]) -> list[TreeCall]:
    """Tous les appels de l'arbre, ordre de parcours en profondeur."""
    out: list[TreeCall] = []
    for block in blocks:
        out.extend(block.calls)
        out.extend(iter_calls(block.blocks))
    return out


EMPTY_TREE: dict[str, object] = {"version": 1, "blocks": []}


TreeFilterGroup.model_rebuild()
TreeBlock.model_rebuild()
